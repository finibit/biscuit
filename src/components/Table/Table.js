import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { themeGet } from '../../themes'
import styles from '../../styles'
import types from '../../types'
import * as parts from './parts'

const CaptionStyled = styled.caption`
  font-family: ${themeGet.fontFamily('primary')};
  font-size: ${themeGet.fontSize(1)};
  font-weight: normal;
  text-align: left;
  padding: ${themeGet.padding(1)} ${themeGet.padding(2)};
`

const TableStyled = styled.table`
  box-sizing: border-box;
  table-layout: fixed;
  border-collapse: collapse;
  font-family: ${themeGet.fontFamily('primary')};
  font-size: ${themeGet.fontSize(1)};
  font-weight: normal;
  padding: 0;
  line-height: ${themeGet.lineHeight('lg')};

  tr {
    border-bottom: ${themeGet.border('light')};
  }

  th, td {
    text-align: left;
    padding: ${themeGet.padding(1)} ${themeGet.padding(2)};
  }

  ${styles.compose(
    styles.border,
    styles.layout
  )};

  ${styles.css};
`

const Table = (props) => {
  const {
    caption,
    children,
    ...rest
  } = props

  return (
    <TableStyled
      {...rest}
    >
      {caption ? (
        <CaptionStyled
          {...rest}
        >
          {caption}
        </CaptionStyled>
      ) : null}
      {children}
    </TableStyled>
  )
}

Table.propTypes = {
  caption: PropTypes.string,
  ...types.border,
  ...types.size,
  ...types.layout,
  ...types.margin
}

Table.Body = parts.TableBody
Table.Cell = parts.TableCell
Table.Footer = parts.TableFooter
Table.Header = parts.TableHeader
Table.HeaderCell = parts.TableHeaderCell
Table.Row = parts.TableRow

export default Table
