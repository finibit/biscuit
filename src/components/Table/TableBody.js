import React from 'react'
import PropTypes from 'prop-types'

const TableBody = (props) => (
  <tbody props={props}>
    {props.children}
  </tbody>
)

TableBody.propTypes = {
  /** Any number of renderable nodes. */
  children: PropTypes.node
}

TableBody.defaultProps = {
  children: null
}

export default TableBody
