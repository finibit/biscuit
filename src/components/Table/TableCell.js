import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const TableCellStyled = styled.td`
	width: ${(props) => props.$width};
	max-width: ${(props) => props.$width};
	margin: 0;
	padding: 0;

	${(props) => props.$css}
`

const TableCell = (props) => {
	const {
		width,
		css,
		...rest
	} = props

	return (
		<TableCellStyled
			$width={width}
			$css={css}
			{...rest}
		/>
	)
}

TableCell.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,

	/** How many columns the cell extends. */
	colSpan: PropTypes.number,

	/** How many rows the cell extends. */
	rowSpan: PropTypes.number,

	/** Fixed width of the cell. */
	width: PropTypes.string,

	/** Custom styles passed to styled-components. */
	css: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.object]),
}

TableCell.defaultProps = {
	children: null,
	colSpan: 1,
	rowSpan: 1,
	width: 'initial',
	css: null,
}

export default TableCell
