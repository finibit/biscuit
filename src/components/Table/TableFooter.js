import React from 'react'
import PropTypes from 'prop-types'

const TableFooter = (props) => (
	<tfoot props={props}>
		{props.children}
	</tfoot>
)

TableFooter.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,
}

TableFooter.defaultProps = {
	children: null,
}

export default TableFooter
