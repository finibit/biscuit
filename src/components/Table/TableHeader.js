import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import styles from '../../themes'

const TableHeaderStyled = styled.thead`
	margin: 0;
	padding: 0;
	${styles.fontFamily};
	${(props) => props.$css}
`

const TableHeader = (props) => {
	const {
		css,
		...rest
	} = props

	return (
		<TableHeaderStyled
			$css={css}
			{...rest}
		/>
	)
}

TableHeader.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,

	/** Custom styles passed to styled-components. */
	css: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.object]),
}

TableHeader.defaultProps = {
	children: null,
	css: null,
}

export default TableHeader
