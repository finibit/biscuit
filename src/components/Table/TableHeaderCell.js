import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const TableHeaderCellStyled = styled.th`
	font-family: ${(props) => props.theme.fontFamily};
	font-weight: normal;
	text-align: left;
	width: ${(props) => props.$width};
	margin: 0;
	padding: 0;

	${(props) => props.$css};
`

const TableHeaderCell = (props) => {
	const {
		colSpan,
		rowSpan,
		width,
		css,
		...rest
	} = props

	return (
		<TableHeaderCellStyled
			colspan={colSpan}
			rowspan={rowSpan}
			$width={width}
			$css={css}
			{...rest}
		/>
	)
}

TableHeaderCell.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,
	/** The scope of the cell. */
	scope: PropTypes.oneOf(['row', 'col', 'rowgroup', 'colgroup', 'auto']),
	/** How many columns the cell extends. */
	colSpan: PropTypes.number,
	/** How many rows the cell extends. */
	rowSpan: PropTypes.number,
	/** Fixed width of the cell. */
	width: PropTypes.string,
	/** Custom styles passed to styled-components. */
	css: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.array]),
}

TableHeaderCell.defaultProps = {
	children: null,
	scope: 'auto',
	colSpan: 1,
	rowSpan: 1,
	width: 'auto',
	css: null,
}

export default TableHeaderCell
