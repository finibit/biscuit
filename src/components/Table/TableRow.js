import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const TableRowStyled = styled.tr`
	margin: 0;
	padding: 0;

	${(props) => props.$css};
`

const TableRow = (props) => {
	const {
		css,
		...rest
	} = props

	return (
		<TableRowStyled
			$css={css}
			{...rest}
		/>
	)
}

TableRow.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,

	/** Custom styles passed to styled-components. */
	css: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.object]),
}

TableRow.defaultProps = {
	children: null,
	css: null,
}

export default TableRow
