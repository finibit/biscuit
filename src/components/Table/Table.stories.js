import React from 'react'
import { storiesOf } from '@storybook/react'
import Wrapper from '../../Wrapper'
import Table from './Table'

const planets = [
  {
    name: 'Mercury',
    volume: '60,827,208,742',
    density: '5.427',
    mass: '330,104,000,000,000,000,000,000'
  },
  {
    name: 'Venus',
    volume: '928,415,345,893',
    density: '5.243',
    mass: '4,867,320,000,000,000,000,000,000'
  },
  {
    name: 'Earth',
    volume: '1,083,206,916,846',
    density: '5.513',
    mass: '5,972,190,000,000,000,000,000,000'
  },
  {
    name: 'Mars',
    volume: '163,115,609,799',
    density: '3.934',
    mass: '641,693,000,000,000,000,000,000'
  },
  {
    name: 'Jupiter',
    volume: '1,431,281,810,739,360',
    density: '1.326',
    mass: '1,898,130,000,000,000,000,000,000,000'
  },
  {
    name: 'Saturn',
    volume: '827,129,915,150,897',
    density: '0.687',
    mass: '568,319,000,000,000,000,000,000,000'
  },
  {
    name: 'Uranus',
    volume: '68,334,355,695,584',
    density: '1.270',
    mass: '86,810,300,000,000,000,000,000,000'
  },
  {
    name: 'Neptune',
    volume: '62,525,703,987,421',
    density: '1.638',
    mass: '102,410,000,000,000,000,000,000,000'
  }
]

storiesOf('Table', module)
  .addDecorator(Wrapper)
  .add('with data', () => (
    <Table caption='Planets of the solar system'>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>Name</Table.HeaderCell>
          <Table.HeaderCell>Volume (km3)</Table.HeaderCell>
          <Table.HeaderCell>Density (g/cm3)</Table.HeaderCell>
          <Table.HeaderCell>Mass (kg)</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {planets.map((planet) => (
          <Table.Row key={planet.name}>
            <Table.Cell>{planet.name}</Table.Cell>
            <Table.Cell>{planet.volume}</Table.Cell>
            <Table.Cell>{planet.density}</Table.Cell>
            <Table.Cell>{planet.mass}</Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>
      <Table.Footer>
        <Table.Row>
          <Table.Cell>Foot 1</Table.Cell>
          <Table.Cell>Foot 2</Table.Cell>
          <Table.Cell>Foot 3</Table.Cell>
          <Table.Cell>Foot 4</Table.Cell>
        </Table.Row>
      </Table.Footer>
    </Table>
  ))
