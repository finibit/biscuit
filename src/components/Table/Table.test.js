import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { css } from 'styled-components'

import Theme from '../Theme'
import Table from '../Table'

describe('Table', () => {
  test('is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('table is rendered with custom css', () => {
    const component = renderer.create((
      <Theme>
        <Table css={{ backgroundColor: '#000' }} />
        <Table css='background-color: #111;' />
        <Table css={css`background-color: #222;`} />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('caption is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table caption='Caption' />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('header is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Header />
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('header with rows is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Header>
            <Table.Row>
              <Table.Cell scope='header' />
              <Table.Cell scope='header' />
            </Table.Row>
            <Table.Row>
              <Table.Cell scope='header' />
              <Table.Cell scope='header' />
            </Table.Row>
          </Table.Header>
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('body is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Body />
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('body with rows is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Body>
            <Table.Row>
              <Table.Cell scope='body' />
              <Table.Cell scope='body' />
            </Table.Row>
            <Table.Row>
              <Table.Cell scope='body' />
              <Table.Cell scope='body' />
            </Table.Row>
          </Table.Body>
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('footer is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Footer />
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('footer with rows is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Table>
          <Table.Footer>
            <Table.Row>
              <Table.Cell scope='footer' />
              <Table.Cell scope='footer' />
            </Table.Row>
          </Table.Footer>
        </Table>
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
