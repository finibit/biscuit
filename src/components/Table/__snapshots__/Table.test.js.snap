// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table body is rendered 1`] = `
.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <tbody
      className="c1"
    />
  </table>
</div>
`;

exports[`Table body with rows is rendered 1`] = `
.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c3 {
  width: initial;
  min-width: initial;
  max-width: initial;
  margin: 0;
  padding: 0;
}

.c2 {
  margin: 0;
  padding: 0;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <tbody
      className="c1"
    >
      <tr
        className="c2"
      >
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="body"
        />
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="body"
        />
      </tr>
      <tr
        className="c2"
      >
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="body"
        />
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="body"
        />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table caption is rendered 1`] = `
.c1 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  font-weight: normal;
  text-align: left;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <caption
      className="c1"
    >
      Caption
    </caption>
  </table>
</div>
`;

exports[`Table footer is rendered 1`] = `
.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <tfoot
      className="c1"
    />
  </table>
</div>
`;

exports[`Table footer with rows is rendered 1`] = `
.c3 {
  width: initial;
  min-width: initial;
  max-width: initial;
  margin: 0;
  padding: 0;
}

.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c2 {
  margin: 0;
  padding: 0;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <tfoot
      className="c1"
    >
      <tr
        className="c2"
      >
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="footer"
        />
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="footer"
        />
      </tr>
    </tfoot>
  </table>
</div>
`;

exports[`Table header is rendered 1`] = `
.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <thead
      className="c1"
    />
  </table>
</div>
`;

exports[`Table header with rows is rendered 1`] = `
.c3 {
  width: initial;
  min-width: initial;
  max-width: initial;
  margin: 0;
  padding: 0;
}

.c1 {
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
}

.c2 {
  margin: 0;
  padding: 0;
}

.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  >
    <thead
      className="c1"
    >
      <tr
        className="c2"
      >
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="header"
        />
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="header"
        />
      </tr>
      <tr
        className="c2"
      >
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="header"
        />
        <td
          className="c3"
          colSpan={1}
          rowSpan={1}
          scope="header"
        />
      </tr>
    </thead>
  </table>
</div>
`;

exports[`Table is rendered 1`] = `
.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
}

<div>
  <table
    className="c0"
  />
</div>
`;

exports[`Table table is rendered with custom css 1`] = `
.c0 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
  background-color: #000;
}

.c1 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
  background-color: #111;
}

.c2 {
  margin: 0;
  table-layout: fixed;
  border-collapse: collapse;
  border: none;
  font-weight: normal;
  width: initial;
  padding: 0;
  background-color: #222;
}

<div>
  <table
    className="c0"
  />
  <table
    className="c1"
  />
  <table
    className="c2"
  />
</div>
`;
